checkQC/qc_checkers/unidentified_index.py

```diff
@@ -0,0 +1,193 @@
+import re
+import functools
+
+from checkQC.qc_checkers.utils import get_complement
+from checkQC.handlers.qc_handler import QCErrorWarning, QCErrorFatal
+
+def unidentified_index(
+    qc_data,
+    significance_threshold,
+    white_listed_indexes=None,
+):
+    """
+    Identify unidentified indices with are significantly overrepresented
+    and suggest possible causes based on the indices found in the samplesheet.
+
+    Possible causes are:
+    - swap between the first and second index
+    - the index is present on another lane
+    - the index complement, reverse or reverse complement is found in the
+    samplesheet.
+
+    Parameters:
+    -----------
+    qc_data: QCData
+        sequencing data to investigate
+    significance_threshold: float
+        threshold above which errors should be reported (in % on the total read
+        counts for that lane)
+    white_listed_indexes: [str]
+        list of regexes. Indices matching this regex will be reported as
+        warnings instead of errors.
+    """
+    samplesheet_matcher = SamplesheetMatcher(qc_data.samplesheet)
+
+    white_listed_indexes = [
+        re.compile(re_index)
+        for re_index in white_listed_indexes
+    ] if white_listed_indexes else []
+
+    qc_errors = []
+    for lane, lane_data in qc_data.sequencing_metrics.items():
+        for barcode in lane_data["top_unknown_barcodes"]:
+            significance = barcode["count"] / lane_data["total_cluster_pf"] * 100.
+            if significance < significance_threshold:
+                continue
+            index = (
+                f"{barcode['index']}+{barcode['index2']}"
+                if barcode.get("index2") else
+                barcode["index"]
+            )
+            is_white_listed = any(
+                re_allowed_index.match(index)
+                for re_allowed_index in white_listed_indexes
+            )
+
+            data = {"index": index, "lane": lane, "causes": []}
+            msg = (
+                f"Overrepresented unknown barcode \"{data['index']}\" on lane {lane} "
+                f"({significance:.1f}% > {significance_threshold:.1f}%)."
+                + (" This barcode is white-listed." if is_white_listed else "")
+            )
+
+            causes = samplesheet_matcher.list_causes(barcode)
```

Comment:
Related to my other comment about storing lane number on the barcode level, I suppose it would be possible to send the lane as another parameter here and use it downstream? I think this solution might be cleaner, just something to consider to avoid having the lane number stored in multiple places.

Reply:
Yes, exactly, either that or add it to the barcode dictionary here.

One drawback with adding a `lane` parameter is that it makes it possible have a mismatch between barcode and lane (i.e. send a barcode from lane 1 while the lane parameter is set to 2).

I'm not sure what's better yet, it seems like either way there is a risk for duplication or mismatch. I need to reflect a little bit more on this.